src/main.rs

```diff
@@ -33,43 +51,122 @@ fn init_logger() {
     env_logger::init()
 }
 
-fn run() -> Fallible<()> {
-    let opt = Opt::from_args();
-    let config = JobConfig::from_file(&opt.jobconfig).context("reading job config")?;
-
+fn gumpi_async(opt: Opt, config: JobConfig) -> impl Future<Item = (), Error = failure::Error> {
+    let progname = config.progname.clone();
+    let cpus_requested = opt.numproc;
     let prov_filter = if opt.providers.is_empty() {
         None
     } else {
         debug!("Chosen providers: {:?}", opt.providers);
         Some(opt.providers)
     };
+    let output_cfg = config.output.clone();
 
-    let mgr = SessionMPI::init(opt.hub, opt.numproc, prov_filter)?;
+    // It's safe to call expect here - at this point opt.jobconfig
+    // is guaranteed to be a valid filepath, which is checked by
+    // JobConfig::from_file in main.rs
+    let sources_dir = opt
+        .jobconfig
+        .parent()
+        .expect("Invalid jobconfig path")
+        .to_owned();
+    let noclean = opt.noclean;
 
-    let deploy_prefix;
-    if let Some(sources) = config.sources {
-        // It's safe to call unwrap here - at this point opt.jobconfig
-        // is guaranteed to be a valid filepath, which is checked by
-        // JobConfig::from_file
-        let prefix = mgr
-            .deploy(opt.jobconfig.parent().unwrap(), &sources, &config.progname)
-            .context("deploying the sources")?;
-        deploy_prefix = Some(prefix);
-    } else {
-        deploy_prefix = None;
-    }
+    SessionMPI::init(opt.hub, prov_filter)
+        .handle_ctrlc()
+        .context("initializing session")
+        .and_then(move |session| {
+            use std::rc::Rc;
+            let session = Rc::new(session);
+            let mut session_clone = Rc::clone(&session);
 
-    mgr.exec(
-        opt.numproc,
-        config.progname,
-        config.args,
-        config.mpiargs,
-        deploy_prefix,
-    )?;
+            info!("available cores: {}", session.total_cpus());
+            let cpus_available = session.total_cpus();
+            if cpus_available < cpus_requested {
+                return Either::A(future::err(format_err!(
+                    "Not enough CPUs available: requested: {}, available: {}",
+                    cpus_requested,
+                    cpus_available
+                )));
+            }
+            info!("Compiling the sources...");
+            let deploy_prefix = if let Some(sources) = config.sources.clone() {
```

Comment:
What does `prefix` mean in this variable name?

Reply:
Asking because it's hard to reason what this variable actually it.

Reply:
Thanks, added a comment.